From 58f6e2861cb37e1a54343867755a6c6bb7e8ca6f Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:10:59 -0400
Subject: [usb] pl2303: clean up line-status handling

Message-id: <1410887463-20674-588-git-send-email-dzickus@redhat.com>
Patchwork-id: 95487
O-Subject: [RHEL7 PATCH 587/591] USB: pl2303: clean up line-status handling
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 6020c3bec359873c1e4081785f220db99694c4e4
Author: Johan Hovold <jhovold@gmail.com>
Date:   Thu Jan 2 22:49:21 2014 +0100

    USB: pl2303: clean up line-status handling

    Clean up line-status handling somewhat.
    Get tty-reference only when actually needed.

    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/serial/pl2303.c b/drivers/usb/serial/pl2303.c
index 0a01fb6..eada524 100644
--- a/drivers/usb/serial/pl2303.c
+++ b/drivers/usb/serial/pl2303.c
@@ -812,7 +812,8 @@ static void pl2303_update_line_status(struct usb_serial_port *port,
  struct tty_struct *tty;
  unsigned long flags;
  unsigned int status_idx = UART_STATE_INDEX;
- u8 prev_line_status;
+ u8 status;
+ u8 delta;
 
  if (spriv->quirks & PL2303_QUIRK_UART_STATE_IDX0)
   status_idx = 0;
@@ -820,24 +821,27 @@ static void pl2303_update_line_status(struct usb_serial_port *port,
  if (actual_length < status_idx + 1)
   return;
 
+ status = data[status_idx];
+
  /* Save off the uart status for others to look at */
  spin_lock_irqsave(&priv->lock, flags);
- prev_line_status = priv->line_status;
- priv->line_status = data[status_idx];
+ delta = priv->line_status ^ status;
+ priv->line_status = status;
  spin_unlock_irqrestore(&priv->lock, flags);
 
- if (priv->line_status & UART_BREAK_ERROR)
+ if (status & UART_BREAK_ERROR)
   usb_serial_handle_break(port);
 
  wake_up_interruptible(&port->port.delta_msr_wait);
 
- tty = tty_port_tty_get(&port->port);
- if (!tty)
-  return;
- if ((priv->line_status ^ prev_line_status) & UART_DCD)
-  usb_serial_handle_dcd_change(port, tty,
-    priv->line_status & UART_DCD);
- tty_kref_put(tty);
+ if (delta & UART_DCD) {
+  tty = tty_port_tty_get(&port->port);
+  if (tty) {
+   usb_serial_handle_dcd_change(port, tty,
+       status & UART_DCD);
+   tty_kref_put(tty);
+  }
+ }
 }
 
 static void pl2303_read_int_callback(struct urb *urb)
-- 
1.7.1